modules/post/windows/gather/enum_snmp.rb

```diff
@@ -3,142 +3,192 @@
-        "Name",
-        "Type"
-      ])
-    print_status("Enumerating community strings")
-    key = "HKLM\\System\\CurrentControlSet\\Services\\SNMP\\Parameters\\ValidCommunities"
-    comm_str = registry_enumvals(key)
-    if not comm_str.nil? and not comm_str.empty?
-      comm_str.each do |c|
-
-        # comm_type is for human display, access_type is passed to the credential
-        # code using labels consistent with the SNMP login scanner
-        case registry_getvaldata(key,c)
-        when 4
-          comm_type = 'READ ONLY'
-          access_type = 'read-only'
-        when 1
-          comm_type = 'DISABLED'
-          access_type = 'disabled'
-        when 2
-          comm_type = 'NOTIFY'
-          access_type = 'notify'
-        when 8
-          comm_type = 'READ & WRITE'
-          access_type = 'read-write'
-        when 16
-          comm_type = 'READ CREATE'
-          access_type = 'read-create'
-        end
-
-        # Save data to table
-        tbl << [c,comm_type]
-
-        register_creds(session.session_host, 161, '', c, 'snmp', access_type)
+        'Name',
+        'Type'
+      ]
+    )
+
+    comm_strings.each do |c|
+      # comm_type is for human display, access_type is passed to the credential
+      # code using labels consistent with the SNMP login scanner
+      type = registry_getvaldata(key, c)
+      case type
+      when 4
+        comm_type = 'READ ONLY'
+        access_type = 'read-only'
+      when 1
+        comm_type = 'DISABLED'
+        access_type = 'disabled'
+      when 2
+        comm_type = 'NOTIFY'
+        access_type = 'notify'
+      when 8
+        comm_type = 'READ & WRITE'
+        access_type = 'read-write'
+      when 16
+        comm_type = 'READ CREATE'
+        access_type = 'read-create'
+      else
+        print_warning("Unknown access type: #{type}")
+        next
       end
-      print_status("")
 
-      # Print table
-      tbl.to_s.each_line do |l|
-        print_status("\t#{l.chomp}")
-      end
-      print_status("")
+      tbl << [c, comm_type]
 
-      # Check who can connect using the Community Strings
-      allowd_for_snmp_query
-
-    else
-      print_error("\tNo Community strings configured")
+      register_creds(session.session_host, 161, '', c, 'snmp', access_type)
+    end
+    print_status
 
+    tbl.to_s.each_line do |l|
+      print_status("\t#{l.chomp}")
     end
+    print_status
+
+    true
   end
 
-  # Method for enumerating the Traps configured
-  def trap_setup
-    print_status("Enumerating Trap Configuration")
-    key = "HKLM\\System\\CurrentControlSet\\Services\\SNMP\\Parameters\\TrapConfiguration"
+  # Enumerate configured SNMP Traps
+  def trap_configuration
+    print_status('Enumerating Trap configuration')
+
+    key = 'HKLM\\System\\CurrentControlSet\\Services\\SNMP\\Parameters\\TrapConfiguration'
+
+    unless registry_key_exist?(key)
+      print_error("\tInsufficient privileges to retrieve SNMP Traps")
+      return
+    end
```

Comment:
This prints an error saying it's a privileges problem when the session is running as SYSTEM and there are no traps configured. That's a bit misleading so let's change the message to something more specific to what's actually going on.


```suggestion
    unless registry_key_exist?(key)
      print_error("\tFailed to retrieve SNMP Traps")
      return
    end
```